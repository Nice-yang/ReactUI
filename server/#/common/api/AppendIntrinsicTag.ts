import { TagCursor } from '../models/file';
import { rendererToMainRequest } from '../utils/ipcCreator';

interface Payload {
  cursor: TagCursor;
  tagName: string;
}

export default rendererToMainRequest<Payload, void>('AppendIntrinsicTag');
