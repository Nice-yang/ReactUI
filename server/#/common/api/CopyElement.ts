import { TagCursor } from '../models/file';
import { rendererToMainRequest } from '../utils/ipcCreator';

interface Payload {
  source: TagCursor;
  target: TagCursor;
}

export default rendererToMainRequest<Payload, void>('CopyElement');
