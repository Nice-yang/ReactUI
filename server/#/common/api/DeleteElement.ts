import { TagCursor } from '../models/file';
import { rendererToMainRequest } from '../utils/ipcCreator';

interface Payload extends TagCursor {}

export default rendererToMainRequest<Payload, void>('DeleteElement');
