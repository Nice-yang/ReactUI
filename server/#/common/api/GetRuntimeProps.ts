import { TagCursor } from '../models/file';
import { mainToRendererRequest } from '../utils/ipcCreator';

export const isReactElementIdentifier = '$$__REACTUI__IS_REACT_ELEMENT';

interface Payload extends TagCursor {}

type Result = Record<string, any>;

export default mainToRendererRequest<Payload, Result>('GetRuntimeProps');
