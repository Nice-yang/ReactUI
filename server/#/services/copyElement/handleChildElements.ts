import { isInterinsicTag } from '#/common/utils';
import { ReplacementBuilder } from '#/utils/ReplacementBuilder';
import { getTagName, traverseNodeReferences } from '#/utils/tsNode';
import { EMPTY } from 'rxjs';
import * as ts from 'typescript';

const handleChildElements = (
  rb: ReplacementBuilder,
  sourceFileDeclarationIdentifiers: ts.BindingName[],
  incrementIdentifierName: (s: string) => string,
) => (elementNode: ts.JsxChild) => {
  if (ts.isJsxElement(elementNode) || ts.isJsxSelfClosingElement(elementNode)) {
    const tags = getTagName(elementNode);
    const openingTag = tags[0];

    if (!isInterinsicTag(openingTag.getText())) {
      tags.forEach(tag => {
        rb.replaceNodeWithText(tag, incrementIdentifierName(tag.getText()));
      });

      // handle props

      return traverseNodeReferences(
        openingTag,
        sourceFileDeclarationIdentifiers,
      );
    }
  }

  return EMPTY;
};

export default handleChildElements;
