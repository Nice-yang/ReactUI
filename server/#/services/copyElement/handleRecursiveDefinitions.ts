import { TagCursor } from '#/common/models/file';
import { getPathRelativeToTarget } from '#/utils/file';
import { ReplacementBuilder } from '#/utils/ReplacementBuilder';
import {
  DeclarationStatement,
  findAncestorNode,
  isJsxLikeElement,
} from '#/utils/tsNode';
import { of } from 'rxjs';
import { tap } from 'rxjs/operators';
import * as ts from 'typescript';

const handleRecursiveDefinitions = (
  sourceCursor: TagCursor,
  targetCursor: TagCursor,
  insertionsMap: Map<DeclarationStatement, ReplacementBuilder>,
  incrementIdentifierName: (s: string) => string,
) => ({
  node,
  nodeDeclaration,
  definition,
  definitionDeclaration,
}: {
  node: ts.Identifier;
  nodeDeclaration: DeclarationStatement;
  definition: ts.BindingName;
  definitionDeclaration: DeclarationStatement;
}) => {
  const newIdentifierName = incrementIdentifierName(node.getText());

  /**
   * handle node
   */
  {
    if (!findAncestorNode(isJsxLikeElement)(node) && nodeDeclaration) {
      if (!insertionsMap.has(nodeDeclaration)) {
        insertionsMap.set(
          nodeDeclaration,
          new ReplacementBuilder(nodeDeclaration),
        );
      }

      insertionsMap
        .get(nodeDeclaration)!
        .replaceNodeWithText(node, newIdentifierName);
    }
  }

  /**
   * handle definition
   */
  {
    if (!insertionsMap.has(definitionDeclaration)) {
      insertionsMap.set(
        definitionDeclaration,
        new ReplacementBuilder(definitionDeclaration),
      );
    }

    let pip$ = of('');

    if (
      ts.isImportDeclaration(definitionDeclaration) &&
      (definitionDeclaration.moduleSpecifier as ts.StringLiteral).text.startsWith(
        '.',
      )
    ) {
      pip$ = getPathRelativeToTarget(
        sourceCursor.fileName,
        targetCursor.fileName,
        (definitionDeclaration.moduleSpecifier as ts.StringLiteral).text,
      ).pipe(
        tap(relativePathToTarget => {
          insertionsMap
            .get(definitionDeclaration)!
            .replaceNodeWithText(
              definitionDeclaration.moduleSpecifier,
              relativePathToTarget,
            );
        }),
      );
    }

    return pip$.pipe(
      tap(() => {
        if (
          ts.isImportSpecifier(definition.parent) &&
          definition.parent.name === definition &&
          !definition.parent.propertyName
        ) {
          insertionsMap
            .get(definitionDeclaration)!
            .replaceNodeWithText(
              definition,
              `${definition.getText()} as ${newIdentifierName}`,
            );
        } else {
          insertionsMap
            .get(definitionDeclaration)!
            .replaceNodeWithText(definition, newIdentifierName);
        }
      }),
    );
  }
};

export default handleRecursiveDefinitions;
