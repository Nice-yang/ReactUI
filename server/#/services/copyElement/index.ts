import { TagCursor } from '#/common/models/file';
import { isInterinsicTag } from '#/common/utils';
import { ReplacementBuilder } from '#/utils/ReplacementBuilder';
import {
  DeclarationStatement,
  findAncestorNode,
  findElementAtCursor$,
  getTagName,
  isDeclarationStatement,
  isJsxLikeElement,
  traverseElement,
  traverseNodeReferences,
} from '#/utils/tsNode';
import { getDeclarationIdentifiersAtSourceFile } from '#/utils/tsNode/getDeclarationIdentifiers';
import incrementIdentifierNameFrom from '#/utils/tsNode/incrementIdentifierNameFrom';
import * as R from 'ramda';
import { EMPTY, forkJoin } from 'rxjs';
import { concatMap, map, switchMap, toArray } from 'rxjs/operators';
import * as ts from 'typescript';

const copyElement$ = (sourceCursor: TagCursor, targetCursor: TagCursor) =>
  forkJoin(
    findElementAtCursor$(sourceCursor),
    findElementAtCursor$(targetCursor),
  ).pipe(
    switchMap(([sourceCursorNode, targetCursorNode]) =>
      forkJoin(
        getDeclarationIdentifiersAtSourceFile(sourceCursorNode.getSourceFile()),
        getDeclarationIdentifiersAtSourceFile(targetCursorNode.getSourceFile()),
      ).pipe(
        switchMap(
          ([
            sourceFileDeclarationIdentifiers,
            targetFileDeclarationIdentifiers,
          ]) => {
            const incrementIdentifierName = incrementIdentifierNameFrom(
              targetFileDeclarationIdentifiers.map(x => x.getText()),
            );

            const rb = new ReplacementBuilder(sourceCursorNode);
            const insertionsMap = new Map<
              DeclarationStatement,
              ReplacementBuilder
            >();

            return traverseElement(sourceCursorNode).pipe(
              concatMap(elementNode => {
                if (
                  ts.isJsxElement(elementNode) ||
                  ts.isJsxSelfClosingElement(elementNode)
                ) {
                  const tags = getTagName(elementNode);
                  const openingTag = tags[0];

                  if (!isInterinsicTag(openingTag.getText())) {
                    tags.forEach(tag => {
                      rb.replaceNodeWithText(
                        tag,
                        incrementIdentifierName(tag.getText()),
                      );
                    });

                    // handle props

                    return traverseNodeReferences(
                      openingTag,
                      sourceFileDeclarationIdentifiers,
                    );
                  }
                }

                return EMPTY;
              }),
              map(({ node, definition }) => {
                return {
                  node,
                  nodeDeclaration: findAncestorNode<DeclarationStatement>(
                    isDeclarationStatement,
                  )(node)!,
                  definition,
                  definitionDeclaration: findAncestorNode<DeclarationStatement>(
                    isDeclarationStatement,
                  )(definition)!,
                };
              }),
              map(
                ({
                  node,
                  nodeDeclaration,
                  definition,
                  definitionDeclaration,
                }) => {
                  const newIdentifierName = incrementIdentifierName(
                    node.getText(),
                  );

                  /**
                   * handle node
                   */
                  {
                    if (
                      !findAncestorNode(isJsxLikeElement)(node) &&
                      nodeDeclaration
                    ) {
                      if (!insertionsMap.has(nodeDeclaration)) {
                        insertionsMap.set(
                          nodeDeclaration,
                          new ReplacementBuilder(nodeDeclaration),
                        );
                      }

                      insertionsMap
                        .get(nodeDeclaration)!
                        .replaceNodeWithText(node, newIdentifierName);
                    }
                  }

                  /**
                   * handle definition
                   */
                  {
                    if (!insertionsMap.has(definitionDeclaration)) {
                      insertionsMap.set(
                        definitionDeclaration,
                        new ReplacementBuilder(definitionDeclaration),
                      );
                    }

                    if (ts.isImportDeclaration(definitionDeclaration)) {
                      // insertionsMap
                      //   .get(definitionDeclaration)!
                      //   .replaceNodeWithText(
                      //     definitionDeclaration.moduleSpecifier,
                      //     getPathRelativeToTarget(
                      //       sourceCursor.fileName,
                      //       targetCursor.fileName,
                      //       definitionDeclaration.moduleSpecifier.getText(),
                      //     ),
                      //   );
                    }

                    if (
                      ts.isImportSpecifier(definition.parent) &&
                      definition.parent.name === definition &&
                      !definition.parent.propertyName
                    ) {
                      insertionsMap
                        .get(definitionDeclaration)!
                        .replaceNodeWithText(
                          definition,
                          `${definition.getText()} as ${newIdentifierName}`,
                        );
                    } else {
                      insertionsMap
                        .get(definitionDeclaration)!
                        .replaceNodeWithText(definition, newIdentifierName);
                    }
                  }
                },
              ),
              toArray(),
              map(() => {
                const insertions = Array.from(insertionsMap.values())
                  .map(a => a.applyReplacements())
                  .reverse();

                return {
                  jsx: rb.applyReplacements(),
                  insertions: R.groupBy(
                    a => (a.startsWith('import') ? 'imports' : 'declarations'),
                    insertions,
                  ),
                };
              }),
            );
          },
        ),
      ),
    ),
  );

export default copyElement$;
