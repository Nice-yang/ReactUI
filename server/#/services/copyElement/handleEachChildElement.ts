import GetRuntimeProps, {
  isReactElementIdentifier,
} from '#/common/api/GetRuntimeProps';
import { isInterinsicTag } from '#/common/utils';
import { ReplacementBuilder } from '#/utils/ReplacementBuilder';
import {
  getAttributes,
  getCursor,
  getTagName,
  traverseNodeReferences,
} from '#/utils/tsNode';
import * as electron from 'electron';
import * as _ from 'lodash';
import { EMPTY } from 'rxjs';
import { switchMap } from 'rxjs/operators';
import * as ts from 'typescript';

const getChildRuntimeValue = (value: any) => {
  let newValue = '';
  switch (true) {
    case value === isReactElementIdentifier:
      newValue = '{null}';
      break;
    case typeof value === 'string':
      newValue = `{"${value}"}`;
      break;
    case typeof value === 'number':
      newValue = `{${value}}`;
      break;
    default:
      newValue = `{${JSON.stringify(value)}}`;
      break;
  }
  return newValue;
};

const handleEachChildElement = (
  rb: ReplacementBuilder,
  sourceFileDeclarationIdentifiers: ts.BindingName[],
  incrementIdentifierName: (s: string) => string,
) => ({
  element: elementNode,
  index,
}: {
  element: ts.JsxChild;
  index: number;
}) => {
  if (ts.isJsxElement(elementNode) || ts.isJsxSelfClosingElement(elementNode)) {
    const tags = getTagName(elementNode);
    const openingTag = tags[0];

    const interinsicTag = isInterinsicTag(openingTag.getText());

    if (!interinsicTag) {
      tags.forEach(tag => {
        rb.replaceNodeWithText(tag, incrementIdentifierName(tag.getText()));
      });
    }

    return GetRuntimeProps.callRenderer(
      electron.BrowserWindow.getFocusedWindow()!,
      getCursor(elementNode),
    ).pipe(
      switchMap(props => {
        const newProps: string[] = [];

        Object.keys(props).forEach(prop => {
          const value = props[prop];

          if (prop === isReactElementIdentifier) {
            return;
          }

          if (prop === 'children') {
            return;
          }

          const newValue = getChildRuntimeValue(value);

          newProps.push(` ${prop}=${newValue}`);
        });

        if (newProps.length) {
          const newPropsText = newProps.join('\n');

          const attributes = getAttributes(elementNode);

          rb.replaceNodeWithText(attributes, newPropsText);
        }

        if (!interinsicTag) {
          return traverseNodeReferences(
            openingTag,
            sourceFileDeclarationIdentifiers,
          );
        }

        return EMPTY;
      }),
    );
  }

  const parent = elementNode.parent;

  if (ts.isJsxElement(parent) || ts.isJsxSelfClosingElement(parent)) {
    return GetRuntimeProps.callRenderer(
      electron.BrowserWindow.getFocusedWindow()!,
      getCursor(parent),
    ).pipe(
      switchMap(props => {
        if (!props.children) {
          return EMPTY;
        }

        if (_.isArray(props.children)) {
          if (props.children[index]) {
            rb.replaceNodeWithText(
              elementNode,
              getChildRuntimeValue(props.children[index]),
            );
          }
        } else {
          rb.replaceNodeWithText(
            elementNode,
            getChildRuntimeValue(props.children),
          );
        }

        return EMPTY;
      }),
    );
  }

  return EMPTY;
};

export default handleEachChildElement;
