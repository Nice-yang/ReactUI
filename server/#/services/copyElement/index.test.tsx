import { isReactElementIdentifier } from '#/common/api/GetRuntimeProps';
import { fixtureFile } from '#/utils/test';
import copyElement$ from './index';

it('should work', done => {
  const sourceFile = fixtureFile(__dirname, 'source1.tsx');
  const targetFile = fixtureFile(__dirname, 'target1.tsx');

  copyElement$(
    { fileName: sourceFile, lineNumber: 16, columnNumber: 3 },
    { fileName: targetFile, lineNumber: 5, columnNumber: 5 },
  ).subscribe(result => {
    expect(result).toMatchSnapshot();

    done();
  });
});

it('handle named imports', done => {
  const sourceFile = fixtureFile(__dirname, 'source2.tsx');
  const targetFile = fixtureFile(__dirname, 'source2.tsx');

  copyElement$(
    { fileName: sourceFile, lineNumber: 6, columnNumber: 5 },
    { fileName: targetFile, lineNumber: 7, columnNumber: 5 },
  ).subscribe(result => {
    expect(result).toMatchSnapshot();

    done();
  });
});

it('handle named as imports', done => {
  const sourceFile = fixtureFile(__dirname, 'source3.tsx');
  const targetFile = fixtureFile(__dirname, 'source3.tsx');

  copyElement$(
    { fileName: sourceFile, lineNumber: 6, columnNumber: 5 },
    { fileName: targetFile, lineNumber: 7, columnNumber: 5 },
  ).subscribe(result => {
    expect(result).toMatchSnapshot();

    done();
  });
});

it('handle imports path', done => {
  const sourceFile = fixtureFile(__dirname, 'source4.tsx');
  const targetFile = fixtureFile(__dirname, 't1/t2/target1.tsx');

  copyElement$(
    { fileName: sourceFile, lineNumber: 5, columnNumber: 5 },
    { fileName: targetFile, lineNumber: 4, columnNumber: 3 },
  ).subscribe(result => {
    expect(result).toMatchSnapshot();

    done();
  });
});

it('copies runtime props', done => {
  const { callRenderer } = require('electron-better-ipc/source/main');

  callRenderer.setMock({
    style: { backgroundColor: 'red' },
    date: 1559042696688,
    value: 'fopobar',
    content: isReactElementIdentifier,
    children: 'foobar',
  });

  const sourceFile = fixtureFile(__dirname, 'props/source1.tsx');
  const targetFile = fixtureFile(__dirname, 'props/target1.tsx');

  copyElement$(
    { fileName: sourceFile, lineNumber: 9, columnNumber: 7 },
    { fileName: targetFile, lineNumber: 4, columnNumber: 10 },
  ).subscribe(result => {
    expect(result).toMatchSnapshot();

    callRenderer.removeMock();

    done();
  });
});

it('copies runtime props with multiple children', done => {
  const { callRenderer } = require('electron-better-ipc/source/main');

  callRenderer.setMock({
    style: { backgroundColor: 'red' },
    date: 1559042696688,
    value: 'fopobar',
    content: isReactElementIdentifier,
    children: ['foobaring', 1],
  });

  const sourceFile = fixtureFile(__dirname, 'props/source2.tsx');
  const targetFile = fixtureFile(__dirname, 'props/target1.tsx');

  copyElement$(
    { fileName: sourceFile, lineNumber: 9, columnNumber: 7 },
    { fileName: targetFile, lineNumber: 4, columnNumber: 10 },
  ).subscribe(result => {
    expect(result).toMatchSnapshot();

    callRenderer.removeMock();

    done();
  });
});

it('copies runtime props with single children', done => {
  const { callRenderer } = require('electron-better-ipc/source/main');

  callRenderer.setMock({
    style: { backgroundColor: 'red' },
    date: 1559042696688,
    value: 'fopobar',
    content: isReactElementIdentifier,
    children: 2,
  });

  const sourceFile = fixtureFile(__dirname, 'props/source3.tsx');
  const targetFile = fixtureFile(__dirname, 'props/target1.tsx');

  copyElement$(
    { fileName: sourceFile, lineNumber: 9, columnNumber: 7 },
    { fileName: targetFile, lineNumber: 4, columnNumber: 10 },
  ).subscribe(result => {
    expect(result).toMatchSnapshot();

    callRenderer.removeMock();

    done();
  });
});

describe('recursive references', () => {
  it('common reference', done => {
    const sourceFile = fixtureFile(__dirname, 'recursive/source1.tsx');
    const targetFile = fixtureFile(__dirname, 'recursive/target1.tsx');

    copyElement$(
      { fileName: sourceFile, lineNumber: 12, columnNumber: 5 },
      { fileName: targetFile, lineNumber: 5, columnNumber: 5 },
    ).subscribe(result => {
      expect(result).toMatchSnapshot();

      done();
    });
  });

  it('check order of reference', done => {
    const sourceFile = fixtureFile(__dirname, 'recursive/source2.tsx');
    const targetFile = fixtureFile(__dirname, 'recursive/target2.tsx');

    copyElement$(
      { fileName: sourceFile, lineNumber: 12, columnNumber: 3 },
      { fileName: targetFile, lineNumber: 5, columnNumber: 5 },
    ).subscribe(result => {
      expect(result).toMatchSnapshot();

      done();
    });
  });
});
