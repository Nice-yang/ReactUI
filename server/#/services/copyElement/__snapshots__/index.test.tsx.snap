// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`copies runtime props 1`] = `
Object {
  "insertions": Object {
    "declarations": Array [
      "const Foo_$ = styled_$.div\`\`;",
    ],
    "imports": Array [
      "import styled_$ from 'styled-components';",
    ],
  },
  "jsx": "<Foo_$ style={{\\"backgroundColor\\":\\"red\\"}}
 date={1559042696688}
 value={\\"fopobar\\"}
 content={null}>{\\"foobar\\"}</Foo_$>",
}
`;

exports[`copies runtime props with multiple children 1`] = `
Object {
  "insertions": Object {
    "declarations": Array [
      "const Foo_$ = styled_$.div\`\`;",
    ],
    "imports": Array [
      "import styled_$ from 'styled-components';",
    ],
  },
  "jsx": "<Foo_$ style={{\\"backgroundColor\\":\\"red\\"}}
 date={1559042696688}
 value={\\"fopobar\\"}
 content={null}>
        {\\"foobaring\\"}{1}
      </Foo_$>",
}
`;

exports[`copies runtime props with single children 1`] = `
Object {
  "insertions": Object {
    "declarations": Array [
      "const Foo_$ = styled_$.div\`\`;",
    ],
    "imports": Array [
      "import styled_$ from 'styled-components';",
    ],
  },
  "jsx": "<Foo_$ style={{\\"backgroundColor\\":\\"red\\"}}
 date={1559042696688}
 value={\\"fopobar\\"}
 content={null}>{2}</Foo_$>",
}
`;

exports[`handle imports path 1`] = `
Object {
  "insertions": Object {
    "imports": Array [
      "import Button_$ from ../../../../../index.ts;",
    ],
  },
  "jsx": "<Button_$>Click Me</Button_$>",
}
`;

exports[`handle named as imports 1`] = `
Object {
  "insertions": Object {
    "imports": Array [
      "import { Button as Button_$1 } from 'Button';",
    ],
  },
  "jsx": "<Button_$1>Click Me</Button_$1>",
}
`;

exports[`handle named imports 1`] = `
Object {
  "insertions": Object {
    "imports": Array [
      "import { Button as Button_$ } from 'Button';",
    ],
  },
  "jsx": "<Button_$>Click Me</Button_$>",
}
`;

exports[`should work 1`] = `
Object {
  "insertions": Object {
    "declarations": Array [
      "const Button_$1 = styled_$1.button\`
  color: red;
\`;",
      "const color_$ = 'blue';",
      "const Button_$ = styled_$.button\`
  color: \${color_$};
\`;",
    ],
    "imports": Array [
      "import * as styled_$1 from 'styled';",
      "import * as styled_$ from 'styled';",
    ],
  },
  "jsx": "<div>
    <p>Source</p>
    <input value=\\"foo\\" />
    <Button_$>Click Me</Button_$>
    <Button_$1>Click Me 2</Button_$1>
  </div>",
}
`;
